### PR TITLE
(PE-37347) add bulk signing function

### PR DESCRIPTION
**Pull request by jonathannewman**
This adds a new function that given a set of csr names, the CA settings and a function to report activity will attempt to sign all the csrs that are valid.  The function returns a map that contains the names of the csrs that were signed, and those that weren't signed.

The function validates alt-names, authorization-extensions, and signatures.

Locking is done across the group to prevent lock contention.  In local testing, it took ~5 seconds to process ~500 records, of which 100 were valid.

### CODE REVIEW

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj [hunk 1]
```diff
@@ -1,34 +1,34 @@
 (ns puppetlabs.puppetserver.certificate-authority
-  (:import (java.io BufferedReader BufferedWriter FileNotFoundException InputStream ByteArrayOutputStream ByteArrayInputStream File Reader StringReader IOException)
+  (:require [clj-time.coerce :as time-coerce]
```
Comment (jonathannewman) [inline src/clj/puppetlabs/puppetserver/certificate_authority.clj:2]: ran the cursive "organize imports" to clean this up.

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj [hunk 2]
```diff
@@ -718,6 +718,29 @@
                   (log/trace (i18n/trs "Finish append to serial file. ")))]
             (ks-file/atomic-write infra-node-serials-path stream-content-fn)))))))
 
+(defn stream-content-to-file
```
Comment (jonathannewman) [inline src/clj/puppetlabs/puppetserver/certificate_authority.clj:721]: Moved this out of "write-cert-to-inventory!" for reuse.

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj [hunk 3]
```diff
@@ -746,33 +769,45 @@
...
-                            (.flush writer)
-                            (log/trace (i18n/trs "Finish append to inventory file. ")))]
+        entry (str formatted-serial-number " " not-before " " not-after " /" subject "\n")]
     (common/with-safe-write-lock inventory-lock inventory-lock-descriptor inventory-lock-timeout-seconds
       (log/debug (i18n/trs "Append \"{1}\" to inventory file {0}" cert-inventory entry))
-      (ks-file/atomic-write cert-inventory stream-content-fn)
+      (ks-file/atomic-write cert-inventory (partial stream-content-to-file cert-inventory entry))
       (maybe-write-to-infra-serial! serial-number cert-name settings))))
 
+(schema/defn ^:always-validate
```
Comment (jonathannewman) [inline src/clj/puppetlabs/puppetserver/certificate_authority.clj:778]: Same as above function, but assumes no locking needed

File: test/unit/puppetlabs/puppetserver/certificate_authority_test.clj [hunk 4]
```diff
@@ -1,34 +1,34 @@
 (ns puppetlabs.puppetserver.certificate-authority-test
-  (:import (java.io StringReader
```
Comment (jonathannewman) [inline test/unit/puppetlabs/puppetserver/certificate_authority_test.clj:2]: Used Cursive to clean up imports.

File: test/unit/puppetlabs/puppetserver/certificate_authority_test.clj [hunk 5]
```diff
@@ -2109,18 +2109,18 @@
         subject (utils/cn "foo")]
     (testing "should not support auto-renewal"
       (is (false? (ca/supports-auto-renewal? (utils/generate-certificate-request keypair subject [] []))))
-      (is (false? (ca/supports-auto-renewal? (utils/generate-certificate-request keypair subject [] [{:oid "1.3.6.1.4.1.34380.1.3.2" :value false}]))))
```
Comment (jonathannewman) [inline test/unit/puppetlabs/puppetserver/certificate_authority_test.clj:2112]: Did a find/replace on that magic string with the constant definition in ca.  Thought it was easier to read.

File: test/unit/puppetlabs/puppetserver/certificate_authority_test.clj [hunk 6]
```diff
@@ -2251,3 +2251,131 @@
     (doseq [i (range 1 10000)]
       (is (= (biginteger i) (ca/base-16-str->biginteger (format "0x%x" i)))))))
 
+(defn generate-csr
```
Comment (jonathannewman) [inline test/unit/puppetlabs/puppetserver/certificate_authority_test.clj:2254]: Helper function to generate a csr for the tests.

File: test/unit/puppetlabs/puppetserver/certificate_authority_test.clj [hunk 7]
```diff
@@ -2251,3 +2251,131 @@
...
+                  :not-signed [(:subject-name csr-info)]}
+                 (ca/sign-multiple-certificate-signing-requests! [(:subject-name csr-info)] settings report-activity)))))
+      (testing "correctly rejects a cert with unapproved extensions"
+        (let [csr-info (generate-csr settings [{:oid "1.9.9.9.9.9.0" :value "true" :critical false}] [{:oid ca/pp_auth_auto_renew-attribute :value true}])]
+          (is (= {:signed     []
+                  :not-signed [(:subject-name csr-info)]}
+                 (ca/sign-multiple-certificate-signing-requests! [(:subject-name csr-info)] settings report-activity))))))
+    (testing "multiple entry with both bad and good csrs"
+      (let [count-range (range 0 100)
+            _ (println "begin generate good csrs")
```
Comment (jonathannewman) [inline test/unit/puppetlabs/puppetserver/certificate_authority_test.clj:2340]: This takes a little while (a few seconds), so I left the print statements in here to make it easier to understand why it is delayed.

**Review by steveax — APPROVED** [review]
👍

**Pull request merged**